### PR TITLE
Ensure chunked responses are cacheable

### PR DESCRIPTION
**Pull request by bdeeney**
Proposed patch with covering test case for #81.

Unfortunately, the newly added [urllib3.response.HTTPResponse.read_chunked()](https://github.com/shazow/urllib3/blob/f21c2a2b73e4256ba2787f8470dbee6872987d2d/urllib3/response.py#L419) method exhausts the response stream by calling `readline()` directly on the CallbackFileWrapper's `fp` attribute.  To ensure the callback is processed after the response stream has been exhausted, this patch registers a response hook on the request.  Not a terribly elegant solution, and I worry about the potential for unintended side effects.  I'd welcome any suggestions for modifying/improving this patch.

### CODE REVIEW

**Comment by bdeeney** [comment]
I've updated this PR after further testing revealed some issues with the previously submitted patch.  This latest version removes the CallbackFileWrapper class entirely, which I think is necessary in order to support caching of server responses sent with `Transfer-Encoding: chunked`.

File: tests/test_chunked_response.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    That's your business. But as-is, it works reasonably well for streaming
+    chunked data from a server.
+    """
+    daemon_threads = True
+
+    @property
+    def base_url(self):
+        return 'http://{addr[0]}:{addr[1]}/'.format(addr=self.server_address)
+
+
```
Comment (ionrock) [inline tests/test_chunked_response.py]: Can we move this to the conftest.py and make sure it gets cleaned up when the test suite ends?

**Comment by ionrock** [comment]
@bdeeney This looks reasonable to me. I'm still a little concerned about the hook bit as I've never used them before and am not clear on how reliable that is compared to explicitly wrapping the file handler. I'm also hoping @dstufft or someone else working on pip could give this a try to see if this breaks them.

**Comment by dstufft** [comment]
I think this is going to end up with the problem that the original callback file wrapper attempted to solve-- namely that we wanted to delay actually caching the response until the calling application itself read all of the data. In pip this is important in order to make the progress reporting work. I do not know for a fact that this would break that though.

**Comment by ionrock** [comment]
@bdeeney So, I've gotten around to testing this out and @dstufft is correct that it breaks the use case where the response body is cached even when the response hasn't been used. I'm working on @rmcgibbo's patch #106 instead.

**Pull request closed without merging**